s3/s3.cc

```diff
@@ -0,0 +1,625 @@
+/*
+ * Copyright (C) 2021-present ScyllaDB
+ */
+
+/*
+ * This file is part of Scylla.
+ *
+ * Scylla is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * Scylla is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Scylla.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "s3/s3.hh"
+#include "to_string.hh"
+#include "utils/phased_barrier.hh"
+#include "utils/memory_data_sink.hh"
+
+#include <seastar/core/reactor.hh>
+#include <seastar/core/circular_buffer.hh>
+#include <seastar/core/sstring.hh>
+#include <seastar/core/gate.hh>
+#include <seastar/util/log.hh>
+#include <seastar/core/coroutine.hh>
+#include <seastar/core/iostream.hh>
+
+#include <unordered_map>
+
+using namespace seastar;
+
+namespace s3 {
+
+seastar::logger log("s3");
+
+struct http_response {
```

Comment:
We already discussed it during the hackathon, so I'll raise the issue again - we already have a `http_response_parser` in Seastar, but it was not complete enough back then (ref: https://github.com/avikivity/scylla/pull/3#discussion_r735143731). I'll verify if it is enough for our purpose now, and if it is, I'll replace it. If it isn't, I suggest (and volunteer) we should just fix the Seastar implementation first, and then use it here.